### PR TITLE
use new project and starterProject git specification 

### PR DESCRIPTION
**Pull request by kadel**
**What type of PR is this?**

/kind api-change
/kind code-refactoring
>
> Documentation changes: Please include [skip ci] in your commit message as well
> /kind documentation
> [skip ci]

**What does does this PR do / why we need it**:
more info in https://github.com/openshift/odo/issues/3862

- ~I needed to fix a lot of tests, and while I was at it I removed some of the 1.0.0 devfile stuff.~ done
- ~This PR requires  https://github.com/openshift/odo/pull/3866 (I'm not able to vendor `"github.com/go-git/go-git/v5"` without go mod)~ done
- ~This PR requires changs in registry https://github.com/odo-devfiles/registry/pull/42~ done

**Which issue(s) this PR fixes**:

Fixes https://github.com/openshift/odo/issues/3862

**PR acceptance criteria**:

- [x] Unit test 

- [x] Integration test 

- [ ] Documentation 

- [x] I have read the [test guidelines](https://github.com/openshift/odo/blob/master/docs/dev/test-architecture.adoc)

**How to test changes / Special notes to the reviewer**:

To test new changes you need to use devfiles from https://github.com/odo-devfiles/registry/pull/42 
This PR includes a temporary change of default registry. On a new systems (like in our CI) this registry will be used.
If you are testing locally you need to either change default registry `odo registry update DefaultDevfileRegistry https://raw.githubusercontent.com/kadel/registry/project-location` or add this as a new registry `odo registry add project-update https://raw.githubusercontent.com/kadel/registry/project-location` and then use `odo create --registry project-update`

### CODE REVIEW

File: pkg/odo/cli/component/create.go
Comment (adisky) [inline pkg/odo/cli/component/create.go:959]: `odo create nodejs --starter`, downloading the starter project but removing the devfile.yaml, `odo create nodejs` creating a devfile.yaml, the folder gets overridden when starter project is downloaded.

**Comment by kadel** [comment]
something went wrong with CI cluster 
```
[oc] error: the server doesn't have a resource type "is"
```
/retest

File: pkg/odo/cli/component/create.go
Comment (kadel) [reply pkg/odo/cli/component/create.go:959]: oh, right. I'm not sure why I didn't notice this before :-/

File: pkg/preference/preference.go
Comment (kadel) [inline pkg/preference/preference.go:85]: This needs to be changed before merging!

**Comment by adisky** [comment]
@kadel Now `odo create` is failing for locally present devfile or devfiles specified with flag `--devfile`

```
$ odo create
Experimental mode is enabled, use at your own risk

Validation
 ✓  Creating a devfile component from devfile path: /home/adisky/test-new-project/test2/devfile.yaml [62577ns]
 ✓  Validating devfile component [93337ns]
 ✗  apiVersion or schemaVersion not present in devfile

```

**Comment by kadel** [comment]
> @kadel Now `odo create` is failing for locally present devfile or devfiles specified with flag `--devfile`

can you share devfile that you are using?

**Comment by adisky** [comment]
> > @kadel Now `odo create` is failing for locally present devfile or devfiles specified with flag `--devfile`
> 
> can you share devfile that you are using?

copied from here https://raw.githubusercontent.com/kadel/registry/project-location/devfiles/nodejs/devfile.yaml in a folder ran `odo create`, to create component from existing devfile and got above error.

**Comment by kadel** [comment]
/hold
Putting on hold so it doesn't get accidentally merged with changes in the registry. 
The https://github.com/odo-devfiles/registry/pull/42 should go in just before this

File: pkg/devfile/parser/parse.go
Comment (mik-dass) [inline pkg/devfile/parser/parse.go:80]: This function should have a parse and validate function like the others e.g ParseFromURLAndValidate()

File: pkg/odo/cli/component/create.go
Comment (adisky) [inline pkg/odo/cli/component/create.go:960]: not sure if am doing something wrong here,  the descriptions says if `revision` not found download default one, It is giving error if specified revision does not exist.
https://github.com/openshift/odo/pull/3867/files#diff-5e7d52b6df7f967077e7b6ab084ca107R186

I tried specifying  valid commit id, branch name, giving error
```
 ✗  couldn't find remote ref "refs/heads/450235d421864f9a1a47a06788be69a321ad63de"
```

am i missing the correct format to specify revisions here?

File: pkg/devfile/parser/parse.go
Comment (kadel) [reply pkg/devfile/parser/parse.go:80]: This is intentional and this is also it doesn't have `Validate` it the name.

File: pkg/odo/cli/component/create.go
Comment (kadel) [reply pkg/odo/cli/component/create.go:960]: To be honest I don't know :-( 
The reference is currently not working. The problem is that ideally, we should know if it is branch, tag, or sha.
I'm thinking about marking "reference" as unsupported for now and try to deal with it in followup PR. What do you think?

File: pkg/odo/cli/component/create.go
Comment (mik-dass) [inline pkg/odo/cli/component/create.go:956]: defer with a spinner end here

File: pkg/odo/cli/component/create.go
Comment (mik-dass) [inline pkg/odo/cli/component/create.go:988]:
```
[mrinaldas@localhost testdir]$ odo create nodejs --starter
Experimental mode is enabled, use at your own risk

Validation
 ✓  Checking devfile existence [16153ns]
 ✓  Checking devfile compatibility [22241ns]
 ✓  Creating a devfile component from registry: project-update [23088ns]
 ✓  Validating devfile component [125229ns]

Starter Project
 ◓  Downloading starter project nodejs...
Please use `odo push` command to create the component with source deployed
```
the download spinner is not ending properly

File: pkg/odo/cli/component/create.go
Comment (adisky) [reply pkg/odo/cli/component/create.go:960]: Agreed with it, may be we log error or show warning that reference is not supported.

File: pkg/odo/cli/component/create.go
Comment (adisky) [inline pkg/odo/cli/component/create.go:952]: are we not supporting `sparseCheckoutDir` for git like projects?

File: pkg/devfile/parser/data/common/types.go
Comment (mik-dass) [inline pkg/devfile/parser/data/common/types.go:184]: the 2.0.0/types.go also needs to  be updated

File: pkg/odo/cli/component/create.go
Comment (kadel) [reply pkg/odo/cli/component/create.go:952]: we should not. This is quite hard to do in git. We can look into it later

File: pkg/odo/cli/component/create.go
Comment (kadel) [reply pkg/odo/cli/component/create.go:989]: fixed

File: pkg/devfile/parser/data/common/types.go
Comment (kadel) [reply pkg/devfile/parser/data/common/types.go:184]: deleted most from 2.0.0/types.go it looks like they are not used anywhere

File: pkg/odo/cli/component/create.go
Comment (kadel) [reply pkg/odo/cli/component/create.go]: Added warning, when revision is set.
If someone figures out how to pass a revision so it is accepted by go-git they will be able to use it and still see the warning.


```
Starter Project
 ⚠  Specifying 'revision' in 'checkoutFrom' is not yet supported in odo.
 ✗  Downloading starter project nodejs-starter from https://github.com/odo-devfiles/nodejs-ex.git [466ms]
 ✗  couldn't find remote ref "refs/heads/e832173192434d38d7a5380152e500c8de96b99e"
```

File: pkg/devfile/parser/parse.go
Comment (mik-dass) [inline pkg/devfile/parser/parse.go:98]: `OverrideEvents()` is unused now and we can remove it along with UpdateEvents() used by it

File: pkg/devfile/parser/parse.go
Comment (mik-dass) [reply pkg/devfile/parser/parse.go:80]: Thanks for adding ParseFromDataAndValidate() :+1:

File: pkg/odo/cli/component/create.go
Comment (mik-dass) [inline pkg/odo/cli/component/create.go:1027]: expected

File: pkg/devfile/parse.go
Comment (mik-dass) [inline pkg/devfile/parse.go:38]: from the given bytes

File: pkg/odo/cli/component/create.go
Comment (mik-dass) [inline pkg/odo/cli/component/create.go:981]: since this is also used in the else if below, can we make it end after the else if ladder?

File: pkg/odo/cli/component/create.go
Comment (kadel) [reply pkg/odo/cli/component/create.go:1027]: this was supposed to be executed ;-)
fixed

File: pkg/devfile/parse.go
Comment (kadel) [reply pkg/devfile/parse.go:38]: fixed

File: pkg/odo/cli/component/create.go
Comment (kadel) [reply pkg/odo/cli/component/create.go:981]: the below spinner has a slightly different message (`remoteUrl` vs `logUrl`), I can't simply reuse it

Comment (mik-dass) [reply pkg/odo/cli/component/create.go:981]: Sorry I thought they were the same spinner

File: pkg/odo/cli/component/create.go
Comment (mik-dass) [inline pkg/odo/cli/component/create.go:977]: happens

File: pkg/odo/cli/component/create.go
Comment (mik-dass) [inline pkg/odo/cli/component/create.go:979]: since the next statement is the same, we don't need this line.

Comment (kadel) [reply pkg/odo/cli/component/create.go]: fixed

File: pkg/odo/cli/component/create.go
Comment (kadel) [reply pkg/odo/cli/component/create.go]: fixed

File: pkg/odo/cli/component/create.go
Comment (adisky) [reply pkg/odo/cli/component/create.go:951]: the openliberty devfile is using it https://github.com/odo-devfiles/registry/blob/master/devfiles/java-openliberty/devfile.yaml#L10 :), need to update this.

**Review by cdrage — APPROVED** [review]
Went through the code, looks great to me.

/lgtm

**Pull request merged**
